pkg/notaryv2/notaryv2.go

```diff
@@ -128,5 +135,226 @@ func (v *notaryV2Verifier) verifyOutcomes(outcomes []*notation.VerificationOutco
+		return &images.Response{Digest: repoDesc.Digest.String(), Statements: statements}, nil
+	}
+
+	return nil, fmt.Errorf("failed to fetch attestations")
+}
+
+func verifyAttestators(ctx context.Context, v *notaryV2Verifier, remoteRepo *remote.Repository, opts images.Options, desc ocispec.Descriptor) (ocispec.Descriptor, error) {
+	if opts.Cert == "" && opts.CertChain == "" {
+		// skips the checks when no attestor is provided
+		v.log.V(2).Info("skipping as no attestators is provided", desc)
+		return desc, nil
+	}
+	certsPEM := combineCerts(opts)
+	certs, err := cryptoutils.LoadCertificatesFromPEM(bytes.NewReader([]byte(certsPEM)))
+	if err != nil {
+		v.log.V(2).Info("failed to parse certificates", "err", err)
+		return ocispec.Descriptor{}, errors.Wrapf(err, "failed to parse certificates")
+	}
+
+	v.log.V(2).Info("parsed certificates")
+	trustStore := NewTrustStore("kyverno", certs)
+	policyDoc := v.buildPolicy()
+	notationVerifier, err := verifier.New(policyDoc, trustStore, nil)
+	if err != nil {
+		v.log.V(2).Info("failed to created verifier", "err", err)
+		return ocispec.Descriptor{}, errors.Wrapf(err, "failed to created verifier")
+	}
+
+	v.log.V(2).Info("created verifier")
+	reference := remoteRepo.Reference.Registry + "/" + remoteRepo.Reference.Repository + "@" + desc.Digest.String()
+	repo, parsedRef, err := parseReference(ctx, reference, opts.RegistryClient)
+	if err != nil {
+		v.log.V(2).Info("failed to parse image reference", "ref", opts.ImageRef, "err", err)
+		return ocispec.Descriptor{}, errors.Wrapf(err, "failed to parse image reference: %s", opts.ImageRef)
+	}
+
+	refer := parsedRef.String()
+	remoteVerifyOptions := notation.RemoteVerifyOptions{
+		ArtifactReference:    refer,
+		MaxSignatureAttempts: 10,
+	}
+
+	v.log.V(2).Info("verification started")
+	targetDesc, outcomes, err := notation.Verify(ctx, notationVerifier, repo, remoteVerifyOptions)
+	if err != nil {
+		v.log.V(2).Info("failed to vefify attestator", "remoteVerifyOptions", remoteVerifyOptions, "repo", repo)
+		return targetDesc, err
+	}
+	if err := v.verifyOutcomes(outcomes); err != nil {
+		return targetDesc, err
+	}
+	if targetDesc.Digest != desc.Digest {
+		v.log.V(2).Info("digest mismatch", "expected", desc.Digest.String(), "found", targetDesc.Digest.String())
+		return targetDesc, errors.Errorf("digest mismatch")
+	}
+	v.log.V(2).Info("attestator verified", "desc", targetDesc.Digest.String())
+
+	return targetDesc, nil
+}
+
+func extractStatements(ctx context.Context, repo *remote.Repository, targetDesc ocispec.Descriptor) ([]map[string]interface{}, error) {
+	statements := make([]map[string]interface{}, 0)
+	data, err := extractStatement(ctx, repo, targetDesc)
+	if err != nil {
+		return nil, err
+	}
+	statements = append(statements, data)
+	return statements, nil
+}
+
+func extractStatement(ctx context.Context, repo *remote.Repository, targetDesc ocispec.Descriptor) (map[string]interface{}, error) {
```

Comment:
Why are there no unit tests for any of the added features? It seems like at least a function like this should be relatively straightforward to write some unit tests for.